src/btree/bt_cursor.c

```diff
@@ -2114,3 +2114,71 @@ __wt_btcur_close(WT_CURSOR_BTREE *cbt, bool lowlevel)
 
     return (ret);
 }
+
+/*
+ * __wt_btcur_bounds_row_position --
+ *     A unpositioned bounded cursor need to start its cursor next and prev walk from the lower or
+ *     upper bound depending on which direction it is going. This function calls cursor row search
+ *     to position the cursor appropriately.
+ */
+int
+__wt_btcur_bounds_row_position(
+  WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, bool next, bool *need_walk)
+{
+
+    WT_CURSOR *cursor;
+    WT_ITEM *bound;
+    bool key_out_of_bounds, valid;
+    uint64_t bound_flag, bound_flag_inclusive;
+
+    key_out_of_bounds = false;
+    cursor = &cbt->iface;
+    bound = next ? &cursor->lower_bound : &cursor->upper_bound;
+    bound_flag = next ? WT_CURSTD_BOUND_UPPER : WT_CURSTD_BOUND_LOWER;
+    bound_flag_inclusive = next ? WT_CURSTD_BOUND_LOWER_INCLUSIVE : WT_CURSTD_BOUND_UPPER_INCLUSIVE;
+
+    if (next)
+        WT_STAT_CONN_DATA_INCR(session, cursor_bounds_next_unpositioned);
+    else
+        WT_STAT_CONN_DATA_INCR(session, cursor_bounds_prev_unpositioned);
+
+    WT_ASSERT(session, WT_DATA_IN_ITEM(bound));
+    __wt_cursor_set_raw_key(cursor, bound);
+    WT_RET(__cursor_row_search(cbt, false, NULL, NULL));
+    /*
+     * If there's an exact match, the row-store search function built the key in the cursor's
+     * temporary buffer.
+     */
+    WT_RET(__wt_cursor_valid(cbt, (cbt->compare == 0 ? cbt->tmp : NULL), WT_RECNO_OOB, &valid));
+
+    /*
+     * Clear the cursor key set flag, as we don't want to return the internal set key to the user.
+     */
+    F_CLR(cursor, WT_CURSTD_KEY_SET);
+    /* If the record is valid, set the cursor's key to the record. */
+    if (valid)
+        WT_RET(__wt_key_return(cbt));
+
+    /*
+     * FIXME-WT-9324: When search_near_bounded is implemented then remove this. If search near
+     * returns a value, ensure it's within the bounds.
+     */
+    if (valid && cbt->compare == 0 && F_ISSET(cursor, bound_flag_inclusive)) {
+        return (0);
+    } else if ((next ? cbt->compare > 0 : cbt->compare < 0) && valid) {
+        /*
+         * If cursor row search returns a non-exact key, check the returned key against the upper
+         * bound if doing a next, and the lower bound if doing a prev to ensure the key is within
+         * bounds. If not, there are no visible records, return WT_NOTFOUND.
+         */
+        if (F_ISSET(cursor, bound_flag)) {
+            WT_RET(__wt_row_compare_bounds(
+              session, cursor, S2BT(session)->collator, next, &key_out_of_bounds));
+            if (key_out_of_bounds)
+                return (WT_NOTFOUND);
+        }
+        return (0);
+    }
+    *need_walk = true;
```

Comment:
Should the code default initialize `*need_walk` to false? Presumably as one of the first things it does - since there are direct returns in the middle of the function.